### PR TITLE
Reveng: Guarding against duplicate or bad metadata and adding warning messages

### PR DESCRIPTION
**Pull request by lajones**
Fix for #3861. We had problems with Foreign Keys or Indexes with same name on different tables. We also experienced some problems where we were receiving NULLs in metadata we expected to be non-null or we were unable to find tables for column metadata or columns for indexes/FKs etc. I've fixed the duplicate issue and now we guard against bad metadata and issue warning messages.

### CODE REVIEW

File: src/EntityFramework.MicrosoftSqlServer.Design/SqlServerDatabaseModelFactory.cs [hunk 1]
```diff
@@ -119,13 +133,28 @@ ELSE null
...
+                    if (!_tables.TryGetValue(TableKey(tableName, schemaName), out table))
+                    {
+                        Logger.LogWarning(
+                            SqlServerDesignStrings.UnableToFindTableForColumn(columnName, schemaName, tableName));
+                        continue;
+                    }
+
                     var dataTypeName = reader.GetString(2);
-                    var nullable = reader.GetBoolean(5);
+                    var nullable = reader.IsDBNull(5) ? false : reader.GetBoolean(5);
```
Comment (natemcmaster) [inline src/EntityFramework.MicrosoftSqlServer.Design/SqlServerDatabaseModelFactory.cs]: This column (`sys.columns.is_nullable`) is never DbNull

File: src/EntityFramework.MicrosoftSqlServer.Design/SqlServerDatabaseModelFactory.cs [hunk 2]
```diff
@@ -119,13 +133,28 @@ ELSE null
...
                     var tableName = reader.GetString(1);
+                    var columnName = reader.GetStringOrNull(3);
                     if (!_tableSelectionSet.Allows(schemaName, tableName))
                     {
                         continue;
                     }
 
+                    if (string.IsNullOrEmpty(columnName))
+                    {
+                        Logger.LogWarning(SqlServerDesignStrings.ColumnNameEmpty(schemaName, tableName));
```
Comment (natemcmaster) [inline src/EntityFramework.MicrosoftSqlServer.Design/SqlServerDatabaseModelFactory.cs]: YAGNI. Columns must have a name.

File: src/EntityFramework.MicrosoftSqlServer.Design/SqlServerDatabaseModelFactory.cs [hunk 3]
```diff
@@ -119,13 +133,28 @@ ELSE null
...
+                        continue;
+                    }
+
+                    TableModel table;
+                    if (!_tables.TryGetValue(TableKey(tableName, schemaName), out table))
```
Comment (natemcmaster) [inline src/EntityFramework.MicrosoftSqlServer.Design/SqlServerDatabaseModelFactory.cs]: This is going to blowup the log output when a user filters a table. We should only log this warning if we hit this unexpectedly i.e. table was allowed by TableSelectionSet AND we didn't find it in `_tables`

File: src/EntityFramework.MicrosoftSqlServer.Design/SqlServerDatabaseModelFactory.cs (hunk 1 above)
Comment (lajones) [inline src/EntityFramework.MicrosoftSqlServer.Design/SqlServerDatabaseModelFactory.cs]: I looked at the definition for the view. It is nullable.

File: src/EntityFramework.MicrosoftSqlServer.Design/SqlServerDatabaseModelFactory.cs [hunk 4]
```diff
@@ -174,41 +202,50 @@ ELSE null
...
+                    var tableName = reader.GetString(1);
+                    var indexName = reader.GetStringOrNull(2);
                     if (!_tableSelectionSet.Allows(schemaName, tableName))
                     {
                         continue;
                     }
 
+                    if (string.IsNullOrEmpty(indexName))
+                    {
+                        Logger.LogWarning(SqlServerDesignStrings.IndexNameEmpty(schemaName, tableName));
```
Comment (natemcmaster) [inline src/EntityFramework.MicrosoftSqlServer.Design/SqlServerDatabaseModelFactory.cs]: Likewise. YAGNI. Indexes must have names.

File: src/EntityFramework.MicrosoftSqlServer.Design/SqlServerDatabaseModelFactory.cs (hunk 2 above)
Comment (lajones) [inline src/EntityFramework.MicrosoftSqlServer.Design/SqlServerDatabaseModelFactory.cs]: I looked at the definition of the view. It is nullable. I agree that's weird and would mean something strange but I'd rather issue a warning than die with the KeyNotFoundException that people are seeing.

File: src/EntityFramework.MicrosoftSqlServer.Design/SqlServerDatabaseModelFactory.cs [hunk 5]
```diff
@@ -216,14 +253,33 @@ private void GetIndexes()
                         {
                             Table = table,
                             Name = indexName,
-                            IsUnique = reader.GetBoolean(3),
-                            IsClustered = (reader.GetString(5) == "CLUSTERED") ? true : default(bool?)
+                            IsUnique = reader.IsDBNull(3) ? false : reader.GetBoolean(3),
```
Comment (natemcmaster) [inline src/EntityFramework.MicrosoftSqlServer.Design/SqlServerDatabaseModelFactory.cs]: is_unique is always defined.

File: src/EntityFramework.MicrosoftSqlServer.Design/SqlServerDatabaseModelFactory.cs (hunk 3 above)
Comment (lajones) [inline src/EntityFramework.MicrosoftSqlServer.Design/SqlServerDatabaseModelFactory.cs]: We do this. See the lines just before this.

File: src/EntityFramework.MicrosoftSqlServer.Design/SqlServerDatabaseModelFactory.cs [hunk 6]
```diff
@@ -216,14 +253,33 @@ private void GetIndexes()
...
                         };
                         table.Indexes.Add(index);
                     }
-                    var columnName = reader.GetString(4);
-                    var column = _tableColumns[ColumnKey(index.Table, columnName)];
-                    index.Columns.Add(column);
+
+                    var columnName = reader.GetStringOrNull(4);
+                    ColumnModel column = null;
+                    if (string.IsNullOrEmpty(columnName))
```
Comment (natemcmaster) [inline src/EntityFramework.MicrosoftSqlServer.Design/SqlServerDatabaseModelFactory.cs]: YAGNI

**Comment by natemcmaster** [comment]
Overall this feels like overdone defensive programming. We have a well-defined, expected content coming from sys.* tables and can program against that expectation.

Also LogWarning seems too high for most of these. LogVerbose instead?

File: src/EntityFramework.MicrosoftSqlServer.Design/SqlServerDatabaseModelFactory.cs (hunk 4 above)
Comment (lajones) [inline src/EntityFramework.MicrosoftSqlServer.Design/SqlServerDatabaseModelFactory.cs]: As above. Weirdly this is not true. In fact this was the case that caused the error the user was seeing. Basically I've guarded against every situation where the appropriate sys view definition allows nulls.

**Comment by lajones** [comment]
As mentioned - the user has found themselves with a DB where these issues do happen. Yes, they are weird schemas and the user has probably made a mistake in creating the schema that way. But I'd much rather issue a warning message than have scaffolding fail with a strange exception. Disagree on the level too - you are probably going to run into problems later on - you need the warning that e.g. the index is missing a column etc.

**Comment by rowanmiller** [comment]
I agree with using the warning level too... as noted the data being returned doesn't really make sense so we should clearly surface that we don't know how to handle it and are just skipping it.

**Comment by natemcmaster** [comment]
I didn't read #3861 before making above comments. Wow, yes that is weird that a user got the database into this state. Nameless indexes and columns? How is that even possible?

But since apparently these edge cases are possible, let's defend. :shipit:

**Comment by lajones** [comment]
Thanks. And yes - I totally get how weird this is. Checked in with commit 454968a.

**Pull request closed without merging**